tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/sasl/SaslScramTest.java

```diff
@@ -0,0 +1,94 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.activemq.artemis.tests.integration.amqp.sasl;
+
+import static org.junit.Assert.assertEquals;
+
+import java.io.File;
+
+import javax.jms.Connection;
+import javax.jms.ConnectionFactory;
+import javax.jms.JMSException;
+import javax.jms.MessageConsumer;
+import javax.jms.MessageProducer;
+import javax.jms.Queue;
+import javax.jms.Session;
+import javax.jms.TextMessage;
+
+import org.apache.activemq.artemis.core.server.embedded.EmbeddedActiveMQ;
+import org.apache.activemq.artemis.spi.core.security.ActiveMQJAASSecurityManager;
+import org.apache.qpid.jms.JmsConnectionFactory;
+import org.apache.qpid.jms.exceptions.JMSSecuritySaslException;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+/**
+ * This test SASL-SCRAM Support
+ */
+public class SaslScramTest {
+
+   private static EmbeddedActiveMQ BROKER;
+
+   @BeforeClass
+   public static void startBroker() throws Exception {
+      String loginConfPath = new File(SaslScramTest.class.getResource("/login.config").toURI()).getAbsolutePath();
+      System.out.println(loginConfPath);
+      System.setProperty("java.security.auth.login.config", loginConfPath);
+      BROKER = new EmbeddedActiveMQ();
+      BROKER.setConfigResourcePath(SaslScramTest.class.getResource("/broker-saslscram.xml").toExternalForm());
+      BROKER.setSecurityManager(new ActiveMQJAASSecurityManager("artemis-sasl-scram"));
+      BROKER.start();
+   }
+
+   @AfterClass
+   public static void shutdownBroker() throws Exception {
+      BROKER.stop();
+   }
+
+   @Test
+   public void testUnencryptedWorksWithAllMechanism() throws JMSException {
+      sendRcv("SCRAM-SHA-1", "hello", "ogre1234");
+      sendRcv("SCRAM-SHA-256", "hello", "ogre1234");
+   }
+
+   @Test(expected = JMSSecuritySaslException.class)
+   public void testEncryptedWorksOnlyWithMechanism() throws JMSException {
+      sendRcv("SCRAM-SHA-1", "test", "test");
+   }
+
+   @Test
+   public void testEncryptedWorksWithMechanism() throws JMSException {
+      sendRcv("SCRAM-SHA-256", "test", "test");
+   }
```

Comment:
The tests names could be more descriptive of what they are doing. I did eventually figure out that 'testEncryptedWorksOnlyWithMechanism' and 'testEncryptedWorksWithMechanism' are 2 paired tests checking that login for the 'test' user only works with the SCRAM-SHA-256 mechanism since the details were presumably generated for that mech and can only work with that (goes back to the comment about indicating how the file contents were created).

That makes me think its a bad idea for it the login module to be able support more than one mechanism. The broker generally shouldnt advertise e.g both SCRAM-SHA1 and SCRAM-SHA-256 (though what is advertised is obviously under the configs control) when using a single module if the credentials for a given user can only work with one SCRAM type when stored they way they should be (obviously with plaintext storage it could do any). I think it might make more sense to have mechanism-specific modules given that.

To that end I would perhaps just skip the SCRAM-SHA-1 support and rename everything for 256. It feels like few should consider using it SHA-1 at this point a newly introduced setup instead of requiring 256.

Reply:
You can have users (e.g. old client) that only support SHA-1 and could generate "legacy entries" for them, while having some never clients use SHA-256 (think about a migration from SHA1- to SHA-256 where not all clients can upgrade) and still advertise support for both mechanism. The client has to choose the mechanism. Also JAAS allows a delegation strategy where several providers are asked in a row and such, so I don't think enforcing/limiting the server is suitable here.

Reply:
Can the file have both types at the same time for a given user? Seems like it cant if its properties based?

Perhaps have separate providers for each mech as I suggested, with their own file, let people who want to have a delegation strategy configure it as such?

Clients should really be unaware of whether the server has their credentials only in SCRAM-SHA-1 or in SCRAM-SHA256 format etc if both mechs are being offered. If a mech is offered its meant to work if the client supports it. Someone with old clients only capable of SCRAM-SHA1 (do you know of any such AMQP 1.0 clients?) if offered, should be able to update their client and simply use the SCRAM-SHA256 if it was offered, and just expect that to work. Thats kind of the point of negotiating what is supported.

Reply:
Theoretical that would be possible if one would save the mechanism used to encrypt the data as well e.g. something like:

username:SHA1=...
username:SHA256=....

I'm just not sure if we should complicate things to much here. If users have complex/advanced deployment demands it might be better to fire off an own LoginModule that is not based on plain properties files.

No one is forced to configure the server to offer more than one mechanism though, if one likes to he should know about the consequences. The scheme is part of the SASL-SCRAM initial setup phase there is no way to "upgrade" older credentials on the server side, so the used mechanism is part of the apriori knowledge of the client as well as his username and password.

Reply:
I dont think that would make sense, so that would be back to seeming sensible to only have each module do one mech point for me.

At the very least, I think the example demonstrates a configuration people shouldnt ever really use it in (plus also having the plain text), so for me its not a great example in that regard. Its obviously reasonable the test covers the plaintext storage variant if its going to be allowed.

Clients shouldnt need apriori knowledge of what mechs the server is storing their details for in general. That would somewhat defeat the reason for SASL being there. Yes, you might need their input in order to introduce a new SCRAM mech support if only storing non-plaintext, thats not quite the same thing.

Reply:
The unitest should at least cover that cases even though its not likely to be used. its not meant as an example for a productive configuration. The same applies for the example itself. It demonstrates the plaintext usage and the encrypted usage I don't see what this should be "bad" in any way.

> Clients shouldnt need apriori knowledge of what mechs the server is storing their details for in general. 

that's how SASL_SCRAM works. this way and there is no way around it. You can only authenticate with a mechanism that was negotiated with the server before, SASL-SCRAM is not meant to "upgrade" or something like that. That's intrinsic to the mechanism itself and nothing that could be negotiated.

Reply:
btw the spec says:

>  For interoperability, all SCRAM clients and servers MUST implement
>    the SCRAM-SHA-1 authentication mechanism, i.e., an authentication
>    mechanism from the SCRAM family that uses the SHA-1 hash function as
>    defined in [RFC3174].

so removing SHA-1 seems to be not an option if we want compliance with the spec.

About different schemes, the spec mandates that

> Authentication information: Information used to verify an identity
>       claimed by a SCRAM client.  The authentication information for a
>       SCRAM identity consists of salt, iteration count, "StoredKey" and
>       "ServerKey" (as defined in the algorithm overview) **for each
>       supported cryptographic hash function**.

So I'll add support for storing multiple encrypted forms for the properties module.

Reply:
I think that probably just betrays its age, when SHA-1 was still widely recommended. The SCRAM-SHA-256 RFC specifically opens talking to it having expected longer longeveity than SHA-1 based mechs, and also later recommends transition to SHA-256. Technically one could implement it (e.g seperately in its own module) without offering it normally and still satisfy that original wording hehe. I'm not agaisnt it being there, though I dont see anyone needing it at this point, removing it was mostly a simple reoute to resolve the multi-types issue and adopt single-mech-per-module semantics in one swoop.

The latter wording doesnt surprise me, its jsut how SASL is meant to work...if you offer mechanisms that means they work, and the client is free to pick whatever one they can/want and it should work.